mlir/lib/Dialect/Vector/Transforms/LowerVectorToFromElementsToShuffleTree.cpp

```diff
@@ -0,0 +1,719 @@
+
+      // For the first level of the tree, permute the vector elements to their
+      // relative position in the final output. For subsequent levels, we
+      // propagate the elements to the next level without permutation.
+      SmallVector<int64_t> shuffleMask;
+      if (currentLevel == 0) {
+        shuffleMask = computePermutationShuffleMask(
+            toElemsDefs[i], lhsInterval, toElemsDefs[i + 1], rhsInterval,
+            fromElemsOp, levelVectorSize);
+      } else {
+        auto lhsShuffleOp = cast<ShuffleOp>(lhsVector.getDefiningOp());
+        auto rhsShuffleOp = cast<ShuffleOp>(rhsVector.getDefiningOp());
+        shuffleMask = computePropagationShuffleMask(lhsShuffleOp, lhsInterval,
+                                                    rhsShuffleOp, rhsInterval,
+                                                    levelVectorSize);
+      }
+
+      Value shuffleVal = rewriter.create<vector::ShuffleOp>(
+          loc, lhsVector, rhsVector, shuffleMask);
+      levelOutputs.push_back(shuffleVal);
+    }
+
+    levelInputs = std::move(levelOutputs);
+    ++currentLevel;
+  }
+
+  assert(levelInputs.size() == 1 && "Should have exactly one result");
+  return levelInputs.front();
+}
+
+/// Gather and unique all the `vector.to_elements` operations that feed the
+/// `vector.from_elements` operation. The `vector.to_elements` operations are
+/// returned in order of appearance in the `vector.from_elements`'s operand
+/// list.
+static LogicalResult
+getToElementsDefiningOps(FromElementsOp fromElemsOp,
+                         SmallVectorImpl<ToElementsOp> &toElemsDefs) {
+  SetVector<ToElementsOp> toElemsDefsSet;
+  for (Value element : fromElemsOp.getElements()) {
+    auto toElemsOp = element.getDefiningOp<ToElementsOp>();
+    if (!toElemsOp)
+      return failure();
+    toElemsDefsSet.insert(toElemsOp);
+  }
+
+  toElemsDefs.assign(toElemsDefsSet.begin(), toElemsDefsSet.end());
+  return success();
+}
+
+/// Pass to rewrite `vector.to_elements` + `vector.from_elements` sequences into
+/// a tree of `vector.shuffle` operations.
+struct ToFromElementsToShuffleTreeRewrite final
+    : OpRewritePattern<vector::FromElementsOp> {
+
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::FromElementsOp fromElemsOp,
+                                PatternRewriter &rewriter) const override {
+    VectorType resultType = fromElemsOp.getType();
+    if (resultType.getRank() != 1)
+      return rewriter.notifyMatchFailure(
+          fromElemsOp, "Multi-dimensional vectors are not supported yet");
+    if (resultType.isScalable())
+      return rewriter.notifyMatchFailure(
+          fromElemsOp,
+          "'vector.from_elements' does not support scalable vectors");
+
+    SmallVector<ToElementsOp> toElemsDefs;
+    if (failed(getToElementsDefiningOps(fromElemsOp, toElemsDefs)))
+      return rewriter.notifyMatchFailure(fromElemsOp, "unsupported sources");
+
+    int64_t numElements =
```

Comment:
It's not clear what `numElements` captures.
```suggestion
    int64_t expectedNumElementsPerVector =
```